src/terminal/adapter/PageManager.hpp

```diff
@@ -0,0 +1,67 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- PageManager.hpp
+
+Abstract:
+- This manages the text buffers required by the VT paging operations.
+--*/
+
+#pragma once
+
+#include "ITerminalApi.hpp"
+#include "til.h"
+
+namespace Microsoft::Console::VirtualTerminal
+{
+    class Page
+    {
+    public:
+        Page(TextBuffer& buffer, const til::rect& viewport, const til::CoordType number) noexcept;
+        TextBuffer& Buffer() const noexcept;
+        til::rect Viewport() const noexcept;
+        til::CoordType Number() const noexcept;
+        Cursor& Cursor() const noexcept;
+        const TextAttribute& Attributes() const noexcept;
+        void SetAttributes(const TextAttribute& attr, ITerminalApi* api = nullptr) const;
+        til::CoordType Top() const noexcept;
+        til::CoordType Bottom() const noexcept;
+        til::CoordType Width() const noexcept;
+        til::CoordType Height() const noexcept;

```

Comment:
I have trouble understanding the difference between these 4 and `Viewport()`. Is there some way to deduplicate this code or give it more descriptive member names? Can we have a single method that returns a `til::rect` instead of these 4?

Reply:
In the VT world, the page represents the mutable area of the buffer, and the viewport (typically referred to as the screen or window) is the visible area of the buffer. The viewport is a subset of the page, and the page is a subset of the buffer.

As currently implemented, our page width matches the buffer width, and our page height matches the viewport height, but that won't always be the case. That's why I wanted to hide those implementation detail in this class.

In terms of naming, these members are always going to be referenced via a `Page` instance with something like `page.Width()` or `page.Height()`, which exactly describes what they represent. The viewport dimensions are now only needed in the `DECRQDE` op to return the visible window size, with `page.Viewport.Width()` and `page.Viewport.Height()`.

I was tempted to add another two methods for that, like `VisibleWidth` and `VisibleHeight`, but that seemed overkill for that one usage. I'm happy to do that if you think it's worthwhile though.

I should also note that the last three commits after `DECRQDE` are almost entirely refactoring that isn't crucial. I tried to keep that separate so it can easily be dropped if you don't like it.

Reply:
After reading your explanation, the `vt100.net` glossary a couple times and stuff like [this](https://vt100.net/docs/vt510-rm/SD.html) I think I finally got it. Pages are really just like our modern buffers with "scrollback", you can move a "viewport" around in it and so on. It's obvious in hindsight but for some reason it had to "click" first. Thanks!

I still feel like Top/Bottom and Width/Height may be expressed as `til::point` and `til::size` respectively, but that's a small nit if anything.

Reply:
> Pages are really just like our modern buffers with "scrollback"

I hate to introduce more confusion, but that's not quite right. They're somewhat similar but different concepts. In the diagram below, `V` is the visible viewport, `V+P` forms the addressable page area (assuming the page is bigger than the viewport - not something we currently support), and `S` represents the scrollback buffer, which contains anything that has scrolled off the top of the page.
```
┌───────┐
│   S   │
│   S   │
│   S   │
├┄┄┄┄┄┄┄┤
│   P   │
╒═══════╕
│   V   │
╘═══════╛
│   P   │
└───────┘
```
Applications only have access to the `V+P` area. If you send a `CUP` home sequence, it'll move the cursor to the top of `P`, not `S`. And you can use sequences like `SD` and `SU` to pan the visible `V` area up and down as far as the boundaries of `P`. But the only way to view the content in `S` is if the user scrolls the viewport there manually.

> I still feel like Top/Bottom and Width/Height may be expressed as til::point and til::size respectively,

I get that it seems neater architecturally, but it just makes things unnecessarily difficult for users of the interface. When 99.99% of the time you only need access to something like `page.width()`, you really don't want to have to keep calling `page.size().width()`. One of the main reasons for this bit of refactoring was to get rid of calls like `textBuffer.GetSize().Dimensions().width`, so if you don't feel strongly about this, I'd really like to keep it as it is.